README.md
# Description

A python implementation of the ETSI DAB EPG standard ([http://pda.etsi.org/pda/home.asp?wki_id=hUkTWWG26f8ABCDFDRWAh ETSI TS 102 818]) incorporating an XML and binary marshaller ([http://pda.etsi.org/pda/home.asp?wki_id=-WO7ND5Wzoz_.%27,-28nGw ETSI TS 102 371]). 

This can be used by broadcasters for producing or parsing EPG data to a DAB multiplex for broadcast, or for general schedule interfacing. It can also be used for producing data based on the DAB EPG standard, such as for [http://www.radioplayer.co.uk UK Radioplayer]

The API closely follows the hierarchy of the XML specification, and hence maps well to the concepts contained within.

# Current status

It is being used in production environments for broadcasters, as well as for testing and evaluation. If you do intend to use it, please bear in mind that issues may well exist and testing should be performed to validate it for your particular use case.

I am in the process of augmenting the library with a range of tests for this purpose.

# TODO

The library generally ignore GI files, and there may be some shortcuts taken when marshalling/unmarshalling, e.g. not supporting DRM systems.

Most importantly, the library is currently missing any profiling considerations in its Binary marshaller - i.e. the splitting of the binary between Basic and Advanced profiles, as well as the combining of the two binary files upon binary unmarshalling.

# Installation

Currently, you'll have to download the source and use it locally. I'm working on making installation easier.

# Usage

## Programme Information

Semantically equivalent to the example given in ETSI TS 102 818 section 4.3.1

```
<?xml version="1.0" encoding="UTF-8"?>
<epg system="DAB">
    <schedule creationTime="2009-02-09T14:52:49.263Z" originator="BBC" version="1">
        <scope startTime="2001-03-01T00:00:00.000Z" stopTime="2001-03-02T18:00:00.000Z">
            <serviceScope id="e1.ce15.c221.0"/>
            <serviceScope id="e1.ce15.c224.0"/>
        </scope>
        <programme id="crid://www.bbc.co.uk;dab/BC81123456a" recommendation="yes" shortId="213456">
            <mediumName xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" xmlns="http://www.worlddab.org/schemas/epg">Gilles Peterson</mediumName>
            <longName xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" xmlns="http://www.worlddab.org/schemas/epg">Gilles Peterson: Worldwide</longName>
            <location>
                <time actualDuration="PT2H" actualTime="2003-12-18T00:00:00.000Z" duration="PT2H" time="2003-12-18T00:00:00.000Z"/>
                <bearer id="e1.ce15.c221.0"/>
            </location>
            <mediaDescription>
                <shortDescription xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" xmlns="http://www.worlddab.org/schemas/epg"><![CDATA[Gilles Peterson brings you two hours of global beats and the best of cool. Including the Worldwide family. KV5 are live from Maida Value with special guests.]]></shortDescription>
            </mediaDescription>
            <mediaDescription/>
            <genre href="urn:tva:metadata:cs:ContentCS:2005:3.6.7">
                <name xmlns="http://www.worlddab.org/schemas/epg"><![CDATA[Rap/Hip Hop/Reggae]]></name>
            </genre>
            <genre href="urn:tva:metadata:cs:ContentCS:2005:3.6.8">
                <name xmlns="http://www.worlddab.org/schemas/epg"><![CDATA[Electronic/Club/Urban/Dance]]></name>
            </genre>
            <genre href="urn:tva:metadata:cs:FormatCS:2005:2.5">
                <name xmlns="http://www.worlddab.org/schemas/epg"><![CDATA[ARTISTIC PERFORMANCE]]></name>
            </genre>
            <genre href="urn:tva:metadata:cs:IntentionCS:2005:1.1">
                <name xmlns="http://www.worlddab.org/schemas/epg"><![CDATA[ENTERTAINMENT]]></name>
            </genre>
            <genre href="urn:tva:metadata:cs:ContentCS:2005:3.6.9">
                <name xmlns="http://www.worlddab.org/schemas/epg"><![CDATA[World/Traditional/Ethnic/Folk music]]></name>
            </genre>
            <memberOf id="crid://www.bbc.co.uk/WorldwideGroup" shortId="1000"/>
            <link description="Email:" xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" url="mailto:gilles.peterson@bbc.co.uk"/>
            <link description="Web:" xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" url="http://www.bbc.co.uk/radio1/urban/peterson"/>
            <keywords xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en"></keywords>
            <programmeEvent shortId="6353">
                <shortName xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" xmlns="http://www.worlddab.org/schemas/epg">Herbert</shortName>
                <mediumName xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" xmlns="http://www.worlddab.org/schemas/epg">Herbert Live</mediumName>
                <longName xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" xmlns="http://www.worlddab.org/schemas/epg">Live session from Herbert</longName>
                <location>
                    <relativeTime duration="PT15M" time="PT45M"/>
                </location>
                <mediaDescription>
                    <shortDescription xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" xmlns="http://www.worlddab.org/schemas/epg"><![CDATA[Live session from Herbert, recorded at Cargo on 24/2/01]]></shortDescription>
                </mediaDescription>
                <mediaDescription/>
            </programmeEvent>
        </programme>
        <programme shortId="59033">
            <mediumName xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en" xmlns="http://www.worlddab.org/schemas/epg">PM</mediumName>
            <location>
                <time duration="PT1H" time="2003-12-18T17:00:00.000Z"/>
                <bearer id="e1.ce15.c224.0"/>
            </location>
            <keywords xmlns:ns0="http://www.w3.org/XML/2001/namespace" ns0:lang="en"></keywords>
        </programme>
    </schedule>
</epg>
```

The code to generate this is:

```
import unittest

from dabepg import *
from dabepg.xml import marshall


class Test(unittest.TestCase):

    def test_build_xml_schedule(self):
        schedule = Schedule(version=2, originator='Global Radio')
        epg = Epg(schedule=schedule)
        
        programme = Programme(213456, crid='crid://bbc.co.uk/4969758988')
        
        programme.names.append(MediumName('Gilles Peterson'))
        programme.names.append(LongName('Gilles Peterson: Worldwide'))
        
        location = Location()
        location.times.append(Time(datetime.datetime(2003, 12, 18, 0, 0, 0, 0), datetime.timedelta(hours=2), actual_time=datetime.datetime(2003, 12, 18, 0, 0, 0, 0), actual_duration=datetime.timedelta(hours=2)))
        location.bearers.append(Bearer('e1.ce15.c221.0'))
        programme.locations.append(location)
        
        programme.media.append(ShortDescription('Gilles Peterson brings you two hours of global beats and the best of cool. Including the Worldwide family. KV5 are live from Maida Value with special guests.'))
        
        programme.genres.append(Genre('urn:tva:metadata:cs:ContentCS:2002:3.6.7', name='Rap/Hip Hop/Reggae'))
        
        programme.memberships.append(Membership(1000, crid='crid://www.bbc.co.uk/WorldwideGroup'))
        
        programme.links.append(Link('mailto:gilles.peterson@bbc.co.uk', description='Email:'))
        
        event1 = ProgrammeEvent(6353, crid='crid://www.bbc.co.uk;dab/BC81123456a', recommendation=True)
        event1.names.append(ShortName('Herbert'))
        event1.names.append(MediumName('Herbert Live'))
        event1.names.append(LongName('Live session from Herbert'))
        event_location = Location(times=[RelativeTime(45 * 60, 15 * 60)])
        event1.locations.append(event_location)
        event1.media.append(ShortDescription('Live session from Herbert, recorded at Cargo on 24/2/01'))
        programme.events.append(event1)
        
        event2 = ProgrammeEvent(59033)
        event2.names.append(MediumName('PM'))
        event2.locations.append(Location(times=[Time(datetime.datetime(2003, 12, 18, 17, 0, 0, 0), datetime.timedelta(hours=1))], bearers=[Bearer('e1.ce15.c224.0')]))
        programme.events.append(event2)
        
        schedule.programmes.append(programme)
        print marshall(epg)
```

## Service Information



debian/changelog
python-dabepg (0.4.0) stable; urgency=low

  * Fixed bug with XML filename generation not printing content ID values to hex
  * Fixed bug with not checking existence of height and width values for an 
    unrestricted logo before XML marshalling
  * Added minimal SI marshalling examples to XML and binary
  * Added PI and SI unmarshalling examples to XML and binary (can read from file or stdin)
  * Made 'id' attribute of Service object a list, to reflect multiple ServiceID elements
  * Fixed bug with Membership marshalling to XML

 -- Ben Poor <ben.poor@thisisglobal.com>  Thu, 25 Jul 2013 15:24:39 +0100

python-dabepg (0.3.0) stable; urgency=low

  * Added ServiceInfo binary marshalling
  * Fixed some silly bugs

 -- Ben Poor <ben.poor@thisisglobal.com>  Fri, 06 Jan 2012 15:03:55 +0000

python-dabepg (0.2.1) stable; urgency=low

  * fix for [9] - serviceScope id attribute has 'id=' in value 
  * fix for [10] - Error in get_scope(): AttributeError: ContentId 
  instance has no attribute 'id' 

 -- Pete Redhead <peteredhead@googlemail.com> Fri, 08 Jul 2011 16:39:12 +0100

python-dabepg (0.2.0) stable; urgency=low

  * fix for [8] - Duplicate serviceScope elements in scope node

 -- Ben Poor <ben.poor@thisisglobal.com>  Wed, 06 Jul 2011 16:38:26 +0100

python-dabepg (0.1.0) testing; urgency=low

  * Initial release

 -- Ben Poor <ben.poor@thisisglobal.com>  Thu, 15 Oct 2010 10:49:00 +0100
 


test/binary.py
import unittest

from dabepg import *
from dabepg.binary import *
import datetime

class CDataTypeTest(unittest.TestCase):

    def test_cdata(self):
        cdata = CData('This is a CDATA test')
        print 'CDATA', cdata, 'encodes as'
        bits = cdata.tobytes()
        print bitarray_to_binary(bits)
        print bitarray_to_hex(bits)
        
class LocationElementTest(unittest.TestCase):
    
    def test_location(self):
        location = Location(times=[RelativeTime(datetime.timedelta(minutes=45), datetime.timedelta(minutes=15))])
        print 'location', location, 'encodes as'
        print bitarray_to_hex(build_location(location).tobytes())
        
class MembershipElementTest(unittest.TestCase):
    
    def test_membership(self):
        membership = Membership(1000, crid='crid://www.bbc.co.uk/WorldwideGroup')
        print 'membership', repr(membership), 'encodes as'
        print bitarray_to_hex(build_membership(membership).tobytes())   
        
class ScopeElementTest(unittest.TestCase):
    
    def test_scope(self):
        start = datetime.datetime(2010, 7, 29, 0, 0, 0, 0)
        end = start + datetime.timedelta(days=1)
        services = [ContentId.fromstring('e1.c181.c2a1.0')]
        scope = Scope(start, end, services)
        print 'scope', scope, 'encodes as'
        bits = build_scope(scope).tobytes()
        print bitarray_to_binary(bits)
        print bitarray_to_hex(bits)
        
class TimepointTypeTest(unittest.TestCase):
    
    def test_encode_utc(self):
        from dabepg.binary import encode_timepoint
        from datetime import datetime
        import pytz
        
        now = datetime(2010, 7, 30, 0, 0, 0, 0)
        now = pytz.timezone('Europe/London').fromutc(now)
        now = now.astimezone(pytz.timezone('Europe/London'))
        bits = encode_timepoint(now)
        print 'timepoint', now, 'encodes as'
        print bitarray_to_binary(bits)
        print bitarray_to_hex(bits)
        
    def test_encode_lto(self):
        from dabepg.binary import encode_timepoint
        from datetime import datetime
        import pytz
        
        now = datetime(2010, 7, 30, 0, 0, 0, 0, pytz.timezone('Europe/London'))
       
        bits = encode_timepoint(now)
        print 'timeepoint', now, 'encodes as'
        print bitarray_to_binary(bits)
        print bitarray_to_hex(bits)
        
class GenreTypeTest(unittest.TestCase):
    
    def test_genre(self):
        from dabepg.binary import encode_genre
        from dabepg import Genre
        
        genre = Genre('urn:tva:metadata:cs:ContentCS:2002:3.6.9', 'World/Traditional/Ethnic/Folk Music')
        print 'genre', genre, 'encodes as'
        bits = encode_genre(genre)
        print bitarray_to_binary(bits)
        print bitarray_to_hex(bits)
        
if __name__ == "__main__":
    unittest.main()

test/build_xml_schedule.py
#===============================================================================
# Python DAB EPG API - Serialize/Deserialize To/From objects to XML/Binary as per
# ETSI specifications TS 102 818 (XML Specification for DAB EPG) and TS 102 
# 371 (Transportation and Binary Encoding Specification for EPG).
# 
# Copyright (C) 2010 Global Radio
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#===============================================================================

import unittest

from dabepg import *
from dabepg.xml import marshall


class Test(unittest.TestCase):


    def test_build_xml_schedule(self):
        schedule = Schedule(version=2, originator='Global Radio')
        epg = Epg(schedule=schedule)
        
        programme = Programme(213456, crid='crid://bbc.co.uk/4969758988')
        
        programme.names.append(MediumName('Gilles Peterson'))
        programme.names.append(LongName('Gilles Peterson: Worldwide'))
        
        location = Location()
        location.times.append(Time(datetime.datetime(2003, 12, 18, 0, 0, 0, 0), datetime.timedelta(hours=2), actual_time=datetime.datetime(2003, 12, 18, 0, 0, 0, 0), actual_duration=datetime.timedelta(hours=2)))
        location.bearers.append(Bearer('e1.ce15.c221.0'))
        programme.locations.append(location)
        
        programme.media.append(ShortDescription('Gilles Peterson brings you two hours of global beats and the best of cool. Including the Worldwide family. KV5 are live from Maida Value with special guests.'))
        
        programme.genres.append(Genre('urn:tva:metadata:cs:ContentCS:2002:3.6.7', name='Rap/Hip Hop/Reggae'))
        
        programme.memberships.append(Membership(1000, crid='crid://www.bbc.co.uk/WorldwideGroup'))
        
        programme.links.append(Link('mailto:gilles.peterson@bbc.co.uk', description='Email:'))
        
        event1 = ProgrammeEvent(6353, crid='crid://www.bbc.co.uk;dab/BC81123456a', recommendation=True)
        event1.names.append(ShortName('Herbert'))
        event1.names.append(MediumName('Herbert Live'))
        event1.names.append(LongName('Live session from Herbert'))
        event_location = Location(times=[RelativeTime(45 * 60, 15 * 60)])
        event1.locations.append(event_location)
        event1.media.append(ShortDescription('Live session from Herbert, recorded at Cargo on 24/2/01'))
        programme.events.append(event1)
        
        event2 = ProgrammeEvent(59033)
        event2.names.append(MediumName('PM'))
        event2.locations.append(Location(times=[Time(datetime.datetime(2003, 12, 18, 17, 0, 0, 0), datetime.timedelta(hours=1))], bearers=[Bearer('e1.ce15.c224.0')]))
        programme.events.append(event2)
        
        schedule.programmes.append(programme)
        print marshall(epg)
    

if __name__ == "__main__":
    #import sys;sys.argv = ['', 'Test.testName']
    unittest.main()

src/dabepg/test/EntityTests.py
import unittest

from dabepg import ContentId

class ContentIdTest(unittest.TestCase):

    def test_parse_contentid(self):
        print ContentId.fromstring('e1.ce15.c221.0.1')
        print ContentId.fromstring('e1.c586')   


if __name__ == "__main__":
    unittest.main()

src/dabepg/tva_genre_2005.py
#===============================================================================
# Python DAB EPG API - Serialize/Deserialize To/From objects to XML/Binary as per
# ETSI specifications TS 102 818 (XML Specification for DAB EPG) and TS 102 
# 371 (Transportation and Binary Encoding Specification for EPG).
# 
# Copyright (C) 2010 Global Radio
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#===============================================================================

from dabepg import Genre

class Intention:
    
    ENTERTAIN = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.1', 'ENTERTAIN')
    INFORM = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.2', 'INFORM')
    Infotainment = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.2.3', 'Infotainment')
    Advice = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.2.4', 'Advice')
    EDUCATE = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.3"', 'EDUCATE')
    ENRICH = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.8', 'ENRICH')
    Inspirational_enrichment = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.8.2', 'Inspirational enrichment')
    


src/dabepg/tva_genre_2009.py
#===============================================================================
# Python DAB EPG API - Serialize/Deserialize To/From objects to XML/Binary as per
# ETSI specifications TS 102 818 (XML Specification for DAB EPG) and TS 102 
# 371 (Transportation and Binary Encoding Specification for EPG).
# 
# Copyright (C) 2010 Global Radio
# 
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
# 
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
# 
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
#===============================================================================

from dabepg import Genre

class Intention:
    
    ENTERTAIN = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.1', 'ENTERTAIN')
    INFORM = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.2', 'INFORM')
    Infotainment = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.2.3', 'Infotainment')
    Advice = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.2.4', 'Advice')
    EDUCATE = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.3"', 'EDUCATE')
    ENRICH = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.8', 'ENRICH')
    Inspirational_enrichment = Genre('urn:tva:metadata:cs:IntentionCS:2005:1.8.2', 'Inspirational enrichment')
    
class Content:
    
    class NonFiction(Genre):
         
        def __init__(self):
             Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1', 'NON-FICTION/INFORMATION')
             
        class News(Genre):
             
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.1', 'News')
                 
    
             Daily_News = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.1', 'Daily news')
             Special_News = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.2', 'Special news/edition')
             Special_Report = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.3', 'Special Report')
             Commentary = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.4', 'Commentary')
             Periodical = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.5', 'Periodical')
             National_Politics = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.6', 'National politics/National assembly')
             Economy = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.7', 'Economy/Market/Financial/Business')
             International = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.8', 'Foreign/International')
             Sports = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.9', 'Sports')
             
             class Cultural(Genre):
                 
                 def __init__(self):
                     Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.1.10', 'Cultural')
                     
                 Arts = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.10.1', 'Arts')
                 Entertainment = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.10.2', 'Entertainment')
                 Film = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.10.3', 'Film')
                 Music = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.10.4', 'Music')
                 Radio = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.10.5', 'Radio')
                 TV = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.10.6', 'TV')
                 
             Local = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.11', 'Local/Regional')
             Traffic = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.12', 'Traffic')
             Weather = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.13', 'Weather forecasts')
             Service_information = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.14', 'Service information')
             Public_affairs = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.15', 'Public affairs')
             Current_affairs = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.16', 'Current affairs')
             Consumer_affairs = Genre('urn:tva:metadata:cs:ContentCS:2009:3.1.1.17', 'Consumer affairs')

        class Religion(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.2', 'Religion/Philosophies')
                 
        class General(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.3', 'General non-fiction')
                 
        class Arts(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.4', 'Arts')     
                 
        class Humanities(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.5', 'Humanities')     
                 
        class Sciences(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.6', 'Sciences')    
                 
        class Human_interest(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.7', 'Human interest')        
                 
        class Transport_and_Communications(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.8', 'Transport and Communications')    
                 
        class Events(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.9', 'Events')                    
                 
        class Media(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.10', 'Media')    
                 
        class Listings(Genre):
         
             def __init__(self):
                 Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.1.11', 'Listings')      
                 
                 
    class Sports(Genre):
     
         def __init__(self):
             Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.2', 'SPORTS')   
                           
             
    class Drama(Genre):
     
         def __init__(self):
             Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.4', 'FICTION/DRAMA') 
             
             
    class Entertainment(Genre):
     
         def __init__(self):
             Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.5', 'AMUSEMENT/ENTERTAINMENT')         
             
             
    class Music(Genre):
     
        def __init__(self):
            Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6', 'Music')                   
             
        class Classical(Genre):
         
            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.1', 'Classical music')
                 
            class Early(Genre):
             
                def __init__(self):
                    Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.1.1', 'Early')        
                    
        class Jazz(Genre):
         
            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.2', 'Jazz')    
                
        class Background(Genre):
         
            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.3', 'Background music')   
                
        class PopRock(Genre):
         
            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.4', 'Pop-rock')      
                
        class Blues_Soul(Genre):
         
            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.5', 'Blues/Rhythm and Blues/Soul/Gospel')
                
        class Country(Genre):
         
            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.6', 'Country and Western')          
                
        class Hiphop(Genre):
         
            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.7', 'Rap/Hip Hop/Reggae')  
                
        class Electronic(Genre):

            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.8', 'Electronic/Club/Urban/Dance')
                
        class World(Genre):
         
            def __init__(self):
                Genre.__init__(self, 'urn:tva:metadata:cs:ContentCS:2009:3.6.9', 'World/Traditional/Ethnic/Folk music')                                                                                                                                                                
             
